### PR TITLE
Simplified the Bitcoin example file, removed the redundancy

### PR DESCRIPTION
**Pull request by NeoZ666**

### CODE REVIEW

File: examples/bitcoin_crate/src/main.rs [hunk 1]
```diff
@@ -100,57 +85,19 @@ fn create_transaction(txinput: Vec<TxIn>, txoutput: Vec<TxOut>) -> Transaction {
...
-    tx.into_iter()
-        .enumerate()
-        .map(|(i, tx)| OutputGroup {
-            value: tx.output.iter().map(|op| op.value.to_sat()).sum(),
-            weight: tx.total_size() as u64,
-            input_count: tx.input.len(),
-            creation_sequence: Some(i as u32),
-        })
-        .collect()
+fn create_outputgroup(tx: Transaction) -> OutputGroup {
```
Comment (delcin-raj) [inline examples/bitcoin_crate/src/main.rs:88]: This function should accept `signed_utxo: TxIn` and `value: u64` as inputs.

You should `map` over `Vec<TxIn>` and transform them to `Vec<OutputGroup>`.

`weight_of(Vec<TxOut>)` should be the input to `calculate_baseweight` function.

We are helping to craft new transaction.

We are not using from existing transaction.

So this example should never demonstrate the creation of the new complete transaction.

**Review by delcin-raj — CHANGES_REQUESTED** [review]
Resolve the misunderstanding.

File: examples/bitcoin_crate/src/main.rs (hunk 1 above)
Comment (delcin-raj) [reply examples/bitcoin_crate/src/main.rs:88]: The value for each `utxo` (TxIn) can be a mock value.
In the comments explain that it has to be queried from bitcoin blockchain for real transactions.

**Comment by NeoZ666** [comment]
We are only using the Outputs for their weights and nothing else. We have hardcoded target values which showcase the usage of coinselection pretty well. 

So we have these options :
a. Remove the hardcoded `target_values` and use the hardcoded `txOut` .
b. Assign a static address as `script_pubkey` for each target value and then calculate base weight for hitting said target values 
c. Just hardcode the `base_weight`.

Third is the fastest and doesn't deal with `txOut` at all. In that case, `Outputs` become redundant and can be removed. 
Second makes it so that we choose the correct base weight everytime.
First is just more lines of code and a bit of change to the `create_select_options` function

**Comment by NeoZ666** [comment]
How it works currently, for reference : 
Screenshot 2024-12-17 173910

File: examples/bitcoin_crate/src/main.rs [hunk 2]
```diff
@@ -1,156 +1,93 @@
...
-        .enumerate()
-        .map(|(i, tx)| OutputGroup {
-            value: tx.output.iter().map(|op| op.value.to_sat()).sum(),
-            weight: tx.total_size() as u64,
-            input_count: tx.input.len(),
-            creation_sequence: Some(i as u32),
-        })
-        .collect()
+/// Creates an OutputGroup from a transaction input and its value
+/// Value would normally be queried from the mainchain for the UTXO being spent
```
Comment (jkciw) [inline examples/bitcoin_crate/src/main.rs:77]:
```suggestion
/// Value would normally be queried from the mainchain for the UTXO being spent
```
/// In practice, the details about the UTXO, used as input, is obtained from the UTXO set maintained by a node.

File: examples/bitcoin_crate/src/main.rs [hunk 3]
```diff
@@ -212,35 +154,23 @@ fn perform_select_coin(utxos: Vec<OutputGroup>, coin_select_options_vec: Vec<Coi
...
-    // Create UTXOs of type OutputGroup to be passed to coin selection
-    let utxos = create_outputgroup(transactions);
-    // Create options for coin selection - modify to pass reference
-    let coin_selection_options = create_select_options(&outputs[0]);
+    // Create coin selection options using output weight
+    let coin_selection_options = create_select_options(outputs[0].weight().to_wu());
 
-    // Perform coin selection
-    perform_select_coin(utxos, coin_selection_options);
+    // Perform coin selection for UTXOs
```
Comment (jkciw) [inline examples/bitcoin_crate/src/main.rs:174]:
```suggestion
    // Perform selection among the available UTXOs
```

**Comment by jkciw** [comment]
My suggestion is, for this example, we can assume that the transaction is a segwit transaction and there will be two outputs. The below mentioned weight values can be hardcoded. 
1. Version - 4 bytes - 16 WU
2. Segwit Marker - 2 bytes - 2 WU
3. No. of Outputs(VarInt) - 1 byte - 4 WU 
4. Value of Output 1 - 8 byte - 32 WU
5. Length of ScriptPubkey 1 (VarInt) - 1 byte - 4 WU
6. Script Pubkey 1 - 22 bytes - 88 WU
7. Value of Output 2 - 8 byte - 32 WU
8. Length of ScriptPubkey 2 (VarInt) - 1 byte - 4 WU
9. Script Pubkey 2 - 22 bytes - 88 WU
10. Lock Time - 4 bytes - 16 WU
11. No. of Witness - 1 byte - 1 WU
**Fields that depend on the inputs chosen :**
a. No. of Inputs - VarInt
b. TxID - 32 bytes - 128 WU - to be multiplied by the total no.of inputs
c. VOut - 4 bytes - 16 WU - to be multiplied by the total no.of inputs
d. Script Sig Lenght - VarInt 
e. Script Sig - VarInt
f. Sequence - 4 bytes - 16 WU - to be multiplied by the total no.of inputs
g. Witness Length - VarInt
h. Witness - VarInt

File: examples/bitcoin_crate/src/main.rs (hunk 3 above)
Comment (NeoZ666) [reply examples/bitcoin_crate/src/main.rs]: Done

File: examples/bitcoin_crate/src/main.rs (hunk 2 above)
Comment (NeoZ666) [reply examples/bitcoin_crate/src/main.rs]: Done

File: examples/bitcoin_crate/src/main.rs [hunk 4]
```diff
@@ -11,236 +11,175 @@ use bitcoin::{
...
-        .and_then(|filedata| json_to_txout(&filedata))
-    {
-        Ok(txout_vec) => txout_vec,
-        Err(e) => {
-            println!("Error reading or parsing outputs: {:?}", e);
-            return;
-        }
-    };
+    // Create a transaction for the purpose of calculating a default weight first
+    let fee = calculate_fee(
```
Comment (delcin-raj) [inline examples/bitcoin_crate/src/main.rs:142]: Move this inside the Ok branch of `perform_coinselection`

File: examples/bitcoin_crate/src/main.rs [hunk 5]
```diff
@@ -11,236 +11,175 @@ use bitcoin::{
...
-    perform_select_coin(utxos, coin_selection_options);
+    // Create OutputGroups from each input
+    let utxos: Vec<OutputGroup> = inputs
+        .clone()
+        .into_iter()
+        .zip(mock_input_values)
+        .map(|(input, value)| OutputGroup {
+            // In practice, the details about the UTXO, used as input, is obtained from the UTXO set maintained by a node.
+            value,
+            weight: input.total_size() as u64,
```
Comment (delcin-raj) [inline examples/bitcoin_crate/src/main.rs:158]: as we discussed `input.segwit_weight()` should be used.

File: examples/bitcoin_crate/src/main.rs [hunk 6]
```diff
@@ -11,236 +11,175 @@ use bitcoin::{
...
+
+    // Pre-calculate transaction weight using a dummy change output
+    // Rationale:
+    // 1. Transaction weight calculation requires both inputs and outputs
+    // 2. We know the target output value (4M sats) but not the change amount yet
+    // 3. Using 0 as placeholder for change output is valid because:
+    //    - Output weight in Bitcoin depends on script size, not value
+    //    - Value field is fixed 8 bytes regardless of amount
+    //    - Only the script pubkey size affects weight
+    let outputs = vec![create_txout(target), create_txout(0)];
```
Comment (delcin-raj) [inline examples/bitcoin_crate/src/main.rs:121]: Remove the `create_txout` function and create two variables,
target_txout and change_txout

File: examples/bitcoin_crate/src/main.rs [hunk 7]
```diff
@@ -11,236 +11,175 @@ use bitcoin::{
...
+    let inputs = create_txins();
```
Comment (delcin-raj) [inline examples/bitcoin_crate/src/main.rs:122]: you can remove the create_txins function as well.

File: examples/bitcoin_crate/src/main.rs [hunk 8]
```diff
@@ -11,236 +11,175 @@ use bitcoin::{
...
+            // In practice, the details about the UTXO, used as input, is obtained from the UTXO set maintained by a node.
+            value,
+            weight: input.total_size() as u64,
+            input_count: 1,
+            creation_sequence: None,
+        })
+        .collect();
+
+    // Perform selection among the available UTXOs, create final transaction
+    match perform_select_coin(&utxos, &coin_selection_option) {
```
Comment (delcin-raj) [inline examples/bitcoin_crate/src/main.rs:165]: log_utxos before this line and directly call `select_coin` function.
No need for this perform_select_coin function.

File: examples/bitcoin_crate/src/main.rs [hunk 9]
```diff
@@ -11,236 +11,175 @@ use bitcoin::{
...
+    // Create coin selection options
+    let coin_selection_option = CoinSelectionOpt {
+        target_value: target,
+        target_feerate: 5.0,
+        long_term_feerate: Some(10.0),
+        min_absolute_fee: 4000,
+        base_weight: calculate_base_weight_btc(
+            outputs.iter().map(|output| output.weight().to_wu()).sum(),
+        ),
+        change_weight: 34,
```
Comment (delcin-raj) [inline examples/bitcoin_crate/src/main.rs:133]: this field should be `change_txout.segwit_weight()`

File: examples/bitcoin_crate/src/main.rs [hunk 10]
```diff
@@ -11,236 +11,175 @@ use bitcoin::{
...
+        change_cost: 8,
```
Comment (delcin-raj) [inline examples/bitcoin_crate/src/main.rs:134]: this should be `calculate_fee(change_cost, long_term_feerate)`

File: examples/bitcoin_crate/src/main.rs [hunk 11]
```diff
@@ -11,236 +11,175 @@ use bitcoin::{
...
+        avg_input_weight: 148,
```
Comment (delcin-raj) [inline examples/bitcoin_crate/src/main.rs:135]: should be calculated from the inputs.

File: examples/bitcoin_crate/src/main.rs [hunk 12]
```diff
@@ -11,236 +11,175 @@ use bitcoin::{
...
+        avg_output_weight: 34,
```
Comment (delcin-raj) [inline examples/bitcoin_crate/src/main.rs:136]: should be the half of target_weight + change_weight

**Review by delcin-raj — CHANGES_REQUESTED** [review]

**Comment by delcin-raj** [comment]
When a user of our lib reads the example, they should not be jumping to function definitions in the example.
So keep a single main function.

File: examples/bitcoin_crate/src/main.rs (hunk 10 above)
Comment (NeoZ666) [reply examples/bitcoin_crate/src/main.rs:134]: done

File: examples/bitcoin_crate/src/main.rs (hunk 4 above)
Comment (NeoZ666) [reply examples/bitcoin_crate/src/main.rs:142]: done

File: examples/bitcoin_crate/src/main.rs (hunk 5 above)
Comment (NeoZ666) [reply examples/bitcoin_crate/src/main.rs:158]: done

File: examples/bitcoin_crate/src/main.rs (hunk 7 above)
Comment (NeoZ666) [reply examples/bitcoin_crate/src/main.rs:122]: done

File: examples/bitcoin_crate/src/main.rs (hunk 8 above)
Comment (NeoZ666) [reply examples/bitcoin_crate/src/main.rs:165]: done

File: examples/bitcoin_crate/src/main.rs (hunk 11 above)
Comment (NeoZ666) [reply examples/bitcoin_crate/src/main.rs:135]: done

File: examples/bitcoin_crate/src/main.rs (hunk 12 above)
Comment (NeoZ666) [reply examples/bitcoin_crate/src/main.rs:136]: done

File: examples/bitcoin_crate/src/main.rs (hunk 6 above)
Comment (NeoZ666) [reply examples/bitcoin_crate/src/main.rs:121]: done

File: examples/bitcoin_crate/src/main.rs (hunk 9 above)
Comment (NeoZ666) [reply examples/bitcoin_crate/src/main.rs:133]: txOut doesn't have a segwit_weight field

Comment (delcin-raj) [reply examples/bitcoin_crate/src/main.rs:127]: Ok, what I meant to say is change_weight should not be hardcoded

**Pull request merged**